fuel-derive/src/attribute.rs

```diff
@@ -0,0 +1,172 @@
+// TODO: nice error messages instead of panics, see: https://stackoverflow.com/a/54394014/2867076
+
+use std::collections::HashMap;
+
+use proc_macro2::{
+    TokenStream,
+    TokenTree,
+};
+use syn::{
+    AttrStyle,
+    Attribute,
+    Meta,
+};
+use synstructure::BindingInfo;
+
+fn parse_attrs(s: &synstructure::Structure) -> HashMap<String, TokenStream> {
+    let mut attrs = HashMap::new();
+
+    for attr in &s.ast().attrs {
+        if attr.style != AttrStyle::Outer {
+            continue
+        }
+        if let Meta::List(ml) = &attr.meta {
+            if ml.path.segments.len() == 1 && ml.path.segments[0].ident == "canonical" {
+                let mut tt = ml.tokens.clone().into_iter().peekable();
+                if let Some(key) = tt.next() {
+                    let key = key.to_string();
+                    if let Some(eq_sign) = tt.peek() {
+                        if eq_sign.to_string() == "=" {
+                            let _ = tt.next();
+                        }
+                    } else {
+                        // Single token, no `=`, so it's a boolean flag.
+                        let old = attrs.insert(key.clone(), TokenStream::new());
+                        if old.is_some() {
+                            panic!("duplicate canonical attribute: {}", key);
+                        }
+                        continue
+                    }
+
+                    // Key-value pair
+                    let value = TokenStream::from_iter(tt);
+                    let old = attrs.insert(key.clone(), value);
+                    if old.is_some() {
+                        panic!("duplicate canonical attribute: {}", key);
+                    }
+                    continue
+                }
+                panic!("enum-level canonical attribute must be a `key = value` pair");
+            }
+        }
+    }
+
+    attrs
+}
+
+/// Pop-level `canonical` attributes for a struct
+pub struct StructAttrs {
+    /// The struct is prefixed with the given byte.
+    /// Useful with`#[canonical(inner_discriminant)]`.
+    pub prefix: Option<TokenStream>,
+}
+
+impl StructAttrs {
+    pub fn parse(s: &synstructure::Structure) -> Self {
+        let mut attrs = parse_attrs(s);
+
+        let prefix = attrs.remove("prefix");
```

Comment:
See field documentation for prefix:

https://github.com/FuelLabs/fuel-vm/blob/e048864f1a4b41e41b2f66f494614a6794e04bc0/fuel-derive/src/attribute.rs#L59-L61

So in practice it's used something like this:

```rust
#[derive(Serialize, Deserialize)]
#[canonical(prefix = 3u64)]
pub struct Prefixed3 {
  a: u64,
  b: u64,
}
```

Now `Prefixed3 { a: 5, b: 7 }` serializes to:

```
00 00 00 00 00 00 00 03 // prefix
00 00 00 00 00 00 00 05 // a
00 00 00 00 00 00 00 07 // b
```

This is especially useful with `#[canonical(inner_discriminant = PrefixType)]`:

```rust
#[derive(Serialize, Deserialize, TryFromPrimitive, IntoPrimitive)]
pub enum Discriminant {
  P1 = 1,
  P2 = 2,
}

#[derive(Serialize, Deserialize)]
#[canonical(prefix = Discriminant::P1)]
pub struct Prefixed1 {
  a: u64,
  b: u64,
}

#[derive(Serialize, Deserialize)]
#[canonical(prefix = Discriminant::P2)]
pub struct Prefixed2 {
  a: u64,
  data: Vec<u8>,
}


#[derive(Serialize, Deserialize]
#[canonical(inner_discriminant = Discriminant)]
pub enum Example {
  P1(Prefixed1),
  P2(Prefixed2),
}
```

Now when serializing `Example`, the discriminant isn't written to output, and instead the inner structs `Prefixed{1,2}` perform the write on serialization. When deserializing, the prefix is read to determine which variant to deserialize, but not consumed by the enum. The variant itself then re-reads the prefix.

This makes `Example` to not take more space than any of it's variants, when it can identify the variant by it's prefix instead. This is similar tricks that can be done with i.e. `union` in `C`. For our purposes this is useful since some structures must always start with a given identifier, and we would still like to warp over them using `enum`. I'm not sure if there's actually any reason those types are prefixed with their identifier, but I'd rather just replicate the old behavior exactly in this PR, and consider any changes in serialization behavior in follow-up PRs.

Reply:
Great, thanks for the clear explanation. Is the prefix variable length or is it required to be a certain length, i.e. byte?

Reply:
It's exactly one word, yes. It could be changed, but since that's the only use case now, I decided to keep it simple.

Reply:
According to the spec, aren't the prefixes generally supposed to be a u8? If word size prefixes matches how we serialized things before, that seems like a bug in the spec implementation.

![image](https://github.com/FuelLabs/fuel-vm/assets/794823/62194fa1-628d-4c20-bb53-ebe555d250a1)

Reply:
Word size prefixes makes sense, given this:
![image](https://github.com/FuelLabs/fuel-vm/assets/794823/850bdf6e-714a-492f-a999-a52bf9fb46c4)